examples/dynamic-lora-sidecar/README

```diff
@@ -0,0 +1,97 @@
+
+## Overview
+
+The sidecar continuously monitors a ConfigMap mounted as a YAML configuration file. This file defines the desired state of LORA adapters, including:
+
+- **Adapter ID:** Unique identifier for the adapter.
+- **Source:** Path to the adapter's source files.
+- **Base Model:** The base model to which the adapter should be applied.
+- **toRemove:** (Optional) Indicates whether the adapter should be unloaded.
+
+The sidecar uses the vLLM server's API to load or unload adapters based on the configuration. It also periodically reconciles the registered adapters on the vLLM server with the desired state defined in the ConfigMap, ensuring consistency.
+
+## Features
+
+- **Dynamic Loading and Unloading:**  Load and unload LORA adapters without restarting the vLLM server.
+- **Continuous Reconciliation:**  Ensures the vLLM server's state matches the desired configuration.
+- **ConfigMap Integration:**  Leverages Kubernetes ConfigMaps for easy configuration management.
+- **Easy Deployment:**  Provides a sample deployment YAML for quick setup.
+
+## Repository Contents
+
+- **`sidecar.py`:**  Python script for the sidecar container.
+- **`Dockerfile`:**  Dockerfile to build the sidecar image.
+- **`configmap.yaml`:**  Example ConfigMap YAML file.
+- **`deployment.yaml`:**  Example Kubernetes deployment YAML.
+
+## Usage
+
+1. **Build the Docker Image:**
+   ```bash
+   docker build -t <your-image-name> .
+2. **Create a configmap:**
+    ```bash
+    kubectl create configmap name-of-your-configmap --from-file=your-file.yaml
+3. **Mount the configmap and configure sidecar in your pod**
+    ```yaml
+     spec:
+      shareProcessNamespace: true
+      containers:
+      - name: inference-server
+        image: vllm/vllm-openai:v0.6.3.post1
+        resources:
+          requests:
+            cpu: 5
+            memory: 20Gi
+            ephemeral-storage: 40Gi
+            nvidia.com/gpu : 1
+          limits:
+            cpu: 5
+            memory: 20Gi
+            ephemeral-storage: 40Gi
+            nvidia.com/gpu : 1
+        command: ["/bin/sh", "-c"]
+        args:
+        - vllm serve meta-llama/Llama-2-7b-hf
+        - --host=0.0.0.0
+        - --port=8000
+        - --tensor-parallel-size=1
+        - --swap-space=16
+        - --gpu-memory-utilization=0.95
+        - --max-model-len=2048
+        - --max-num-batched-tokens=4096
+        - --disable-log-stats
+        - --enable-loras
+        - --max-loras=5
+        env:
+        - name: DEPLOY_SOURCE
+          value: UI_NATIVE_MODEL
+        - name: MODEL_ID
+          value: "Llama2-7B"
+        - name: AIP_STORAGE_URI
+          value: "gs://vertex-model-garden-public-us/llama2/llama2-7b-hf"
+        - name: VLLM_ALLOW_RUNTIME_LORA_UPDATING
+          value: "true"
+        volumeMounts:
+        - mountPath: /dev/shm
+          name: dshm
+      initContainers:
```

Comment:
This is an init container which runs prior to the main container loading. However, the container calls APIs of the main container (which is not yet up). Does this create any risk? 

See logs:

```sh
configmap-reader-1 2024-10-25 12:48:12,713 - ERROR - Error communicating with vLLM server: HTTPConnectionPool(host='localhost', port=8000): Max retries exceeded with url: /v1/models (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f8ec15f1690>: Failed to establish a new connection: [Errno 111] Connection refused'))
configmap-reader-1 2024-10-25 12:48:12,714 - ERROR - Error communicating with vLLM server: HTTPConnectionPool(host='localhost', port=8000): Max retries exceeded with url: /v1/models (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f8ec15f1fc0>: Failed to establish a new connection: [Errno 111] Connection refused'))
```

Reply:
Yes, good point how to coordinate , can the sidecar call the kube apiserver, list pods and look at container status and we recommend customers to configure readiness probe to hit the /health endpoint of server. Or best way here is just to hit /health endpoint from sidecar for until server is healthy

Reply:
Added check to see if server is healthy